yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -253,10 +269,26 @@ private[yarn] class YarnAllocator(
    * added in recent versions.
    */
   private def createContainerRequest(resource: Resource): ContainerRequest = {
+    // filter nodes which don't have containers but node preference is required.
+    val nodes = {
+      val unusedNodes = preferredNodeLocationToUses.filterKeys(_ == false).keys
+      if (unusedNodes.isEmpty) {
+        null
+      } else {
+        unusedNodes.toArray
+      }
+    }
+
+    val racks = if (preferredRackLocations.isEmpty) {
+      null
+    } else {
+      preferredRackLocations.toArray
+    }
+
     nodeLabelConstructor.map { constructor =>
-      constructor.newInstance(resource, null, null, RM_REQUEST_PRIORITY, true: java.lang.Boolean,
+      constructor.newInstance(resource, nodes, racks, RM_REQUEST_PRIORITY, true: java.lang.Boolean,
```

Comment:
YARN blows up if you you try to create requests with "relax=true" and "relax=false" on the same priority, or if you try and set relax=true on a request with locality requirements. If that's not surfaced yet is possibly just luck as requests are being satisfied fast enough the conflict  hasn't surfaced.

I'd recommend
1. only set that relax bit if the nodes or racks lists are non-null
2. use a separate priority for relaxed vs unrelaxed requests.